### PR TITLE
Deepseek-R1 GGUF support

### PR DESCRIPTION
**Pull request by SandroPats**
## Motivation

Support model unsloth/DeepSeek-R1-GGUF (https://github.com/sgl-project/sglang/issues/3973)

## Modifications

Adding support of unsloth/DeepSeek-R1-GGUF models. Reused ideas and some code from similar pr from vllm (https://github.com/vllm-project/vllm/pull/13167). Current design assumes you provide config.json from [Unsloth GGUF repo](https://huggingface.co/unsloth/DeepSeek-R1-GGUF/tree/main) and tokenizer.json, configuration.py and modeling.py from [Original DeepSeek-R1 hf repo](https://huggingface.co/deepseek-ai/DeepSeek-R1/tree/main). Config can be provided through the new server argument: --hf-config-path

Current implementation heavily relies on vllm gguf modules. Unfortunately those modules are not compatible with cuda graphs, so it should be disabled: --disable-cuda-graph.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/models/deepseek_v2.py [hunk 1]
```diff
@@ -730,6 +744,8 @@ def _dispatch_mla_subtype():
             else:
                 return AttnForwardMethod.MLA
 
+        #if self.quant_config.get_name() == 'gguf':
+        #    return AttnForwardMethod.MHA
```
[the remark this answers is not in the document]
Comment (SandroPats) [reply python/sglang/srt/models/deepseek_v2.py]: Did this for debugging, currently MLA works just fine. I will remove this

File: python/sglang/srt/models/deepseek_v2.py [hunk 2]
```diff
@@ -2131,7 +2176,7 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]], is_nextn=Fal
                         )
                         weight_loader(param, loaded_weight)
 
-        self.post_load_weights(is_nextn=is_nextn, weight_names=weight_names)
+        self.post_load_weights(is_nextn=is_nextn)
```
[the remark this answers is not in the document]
Comment (SandroPats) [reply python/sglang/srt/models/deepseek_v2.py]: Just wrong conflict resolution, I'll revert this

File: python/sglang/srt/models/deepseek_v2.py [hunk 3]
```diff
@@ -860,15 +876,20 @@ def forward_normal_prepare(
...
-                [self.q_lora_rank, self.kv_lora_rank + self.qk_rope_head_dim], dim=-1
-            )
+            if self.fuse_qkv_a_proj:
+                q, latent_cache = self.fused_qkv_a_proj_with_mqa(hidden_states)[0].split(
+                    [self.q_lora_rank, self.kv_lora_rank + self.qk_rope_head_dim], dim=-1
+                )
+            else:
+                q_c = self.q_a_proj(hidden_states)[0]
+                q = self.q_a_layernorm(q_c)
             q = self.q_a_layernorm(q)
```
[the remark this answers is not in the document]
Comment (SandroPats) [reply python/sglang/srt/models/deepseek_v2.py:889]: No, not intentional. Thanks for pointing this out

File: python/sglang/srt/model_loader/loader.py [hunk 4]
```diff
@@ -1213,9 +1213,28 @@ def _get_gguf_weights_map(self, model_config: ModelConfig):
...
+                )
+                gguf_to_hf_name_map[f"blk.{idx}.ffn_down_exps.weight"] = (
+                    f"model.layers.{idx}.mlp.experts.0.down_proj.weight"
+                )
+                gguf_to_hf_name_map[f"blk.{idx}.ffn_gate_exps.weight"] = (
+                    f"model.layers.{idx}.mlp.experts.0.gate_proj.weight"
+                )
+                gguf_to_hf_name_map[f"blk.{idx}.ffn_up_exps.weight"] = (
+                    f"model.layers.{idx}.mlp.experts.0.up_proj.weight"
+                )
```
[the remark this answers is not in the document]
Comment (SandroPats) [reply python/sglang/srt/model_loader/loader.py:1236]: Quantized gguf weights are stored in way that is incompatible with torch.stack (dequantization result will be different)
